### ISSUE TITLE
CKEditor: Editor does not display on formatted text fields requiring different text formats

### ISSUE THREAD

**Issue by sethhill** [backdrop/backdrop-issues#6771]
## Description of the bug

This appears to happen where there are two formatted text fields on the same content type. For example, if the first field added allows only the format "Limited HTML," and then a second field whose allowed format is "Filtered HTML," CKEditor will not appear on the second field. If you change both fields to "Limited HTML," the editor appears on both.

## Steps To Reproduce

To reproduce the behavior:
1. Create a content type with one formatted text field. Allow one text format that utilizes CKEditor.
2. View the edit screen, the editor should appear.
3. Add aa second formatted text field on that content type, but allow a different text format that utilizes CKEditor.
4. View the edit screen. The editor should appear on the first field, but not on the second.

## Actual behavior

CKEditor does not consistently appear on formatted text fields that allow different formats from one another.

## Expected behavior

CKEditor should appear on all fields if configured, and display the correct editor settings to match the text format.

**Comment by indigoxela** [comment]
I can confirm the bug following the steps. Might be a glitch in the Filter module, but I'm not sure, yet.

A screenshot of what happens:

![no-editor](https://github.com/user-attachments/assets/15e3047e-cad5-48ee-a889-d39e6539ca77)

**Comment by indigoxela** [comment]
LOL (OK, not funny, it's clearly a bug)... if one switches the order of fields, so the one with the editor comes AFTER the plain text one in the node form, the editor initializes properly.

This must be a bug with the Filter module (also tested with TinyMCE, same result, so no prob with CKEditor).

And my current assumption is, that in some loop or array merge the latter (plain text) format overrides the settings for the editor format.
It's just... like always when trying to debug the filter module - it's a complicated beast.

**Comment by indigoxela** [comment]
As I suspected, it's a problem of code order. Switching the order seems to fix the problem, that a later filter format without settings wipes out the settings of the previous one.

I belief, this problem also deserves a functional test (TBD).

And - where are my manners :grin: - welcome @sethhill to the Backdrop issue queue and many thanks for reporting this bug.

**Comment by quicksketch** [comment]
Thank you @sethhill for the bug report!

@indigoxela Thanks for the fix. This doesn't seem quite right to me. If `#attached` is getting set on two different fields, shouldn't the options get combined and both fields work, regardless of which one was defined first? The current PR might be good enough as-is, but I feel like there's something deeper here that causes the current `#attached` properties not to work right.

**Comment by indigoxela** [comment]
> The current PR might be good enough as-is, but I feel like there's something deeper here...

Agreed. It's really odd, how the Filter module handles this.

Note that the libraries got attached correctly in filter_get_attached(), only the js settings were reset to empty array, if a later field has no format with js settings. The actual problem is somewhere deeper (should this even run multiple times per page, for every field independently...), but I suspect, that requires a bigger rewrite.
The (logic) problem got exposed by allowing formats to get limited per field (suppress formats), but IMO existed before.

**Comment by quicksketch** [comment]
> the js settings were reset to empty array, if a later field has no format with js settings

Hm, I think what's supposed to happen here is there that the settings are supposed to passed through `backdrop_array_merge_deep()` and get combined, with the later keys taking any precedent over the previous ones.

But looking at `backdrop_add_js()`, I see this segment:

```
        // All JavaScript settings are placed in the header of the page with
        // the library weight so that inline scripts appear afterwards.
        if (isset($options['key'])) {
          $javascript['settings']['data'][$options['key']] = $data;
        }
```

`filter_get_attached()` has the following:

```
    $attached['js'][] = array(
      'type' => 'setting',
      'key' => 'filter_formats',
      'data' => array('filter' => array('formats' => $settings)),
    );
```

So since `key` in `filter_get_attached()` is specified, it overwrites the previous value in `backdrop_add_js()`, rather than merging them.

I see two possible solutions (haven't tested either): 1. Simply remove the `key` from `filter_get_attached()`. 2. Change that line in `backdrop_add_js()` to merge the values rather than overwriting them.

**Comment by indigoxela** [comment]
I think, we agree, that the Filter module doesn't do that in a smart way.

However, we can't change the js settings (json) array structure, as this would affect all modules that rely on the current one.

And it seems weird to fix the problem caused by Filter in backdrop_add_js(), which might break existing functionality elsewhere.

My current favorite solution's still to simply switch the code order to prevent stub or completely empty settings. To me this seems an approach without side effects. Of am I missing something?

**Comment by quicksketch** [comment]
Hm, on further consideration I think you're right @indigoxela. Let's just stick with what works and document it in the code. Let's go with your solution.

For test coverage: it's a little tough because this is JS behavior we need to verify. Perhaps we simply check if the editor configuration for all text formats is present even when a different text format is used? There is probably already an example in the tests somewhere of verifying the contents of the `Backdrop` JSON object.

**Comment by indigoxela** [comment]
> For test coverage: it's a little tough because this is JS behavior we need to verify

I ended up with string comparison on script tag content... :stuck_out_tongue_winking_eye: 

Functional test added. :smile:

**Comment by laryn** [comment]
I have tested and this change does seem to do the job. Here's what I tested:

- In the sandbox, I created a new "Simple" text format with a scaled down editor experience
- I added a new long text field to the Page content type, which only allowed "Simple" format
- I set the default format for the body on the Page content type to be the default "Basic" format
- Checked when creating Page content that both formats display correctly.
![CleanShot 2024-12-13 at 08 56 48@2x](https://github.com/user-attachments/assets/3c941ac7-f91a-40da-9d10-69747dd06e82)
- Exported full configuration from the sandbox (attached below if anyone wants to grab those configs for testing) and imported to a new fresh site.
- Checked when creating Page content on the fresh site **without this PR**, the text formats do not display correctly:
![CleanShot 2024-12-13 at 09 02 10@2x](https://github.com/user-attachments/assets/cd7c070b-61f1-444b-81f5-bd8438ce6b6c)

Here's the quick config changes if anyone wants to test on a local sandbox quickly:
[config-issue-6771.tar.gz](https://github.com/user-attachments/files/18127901/config-issue-6771.tar.gz)

**Comment by laryn** [comment]
Spot checking those screenshots: did the field description move from above the field to below the field in this PR?

**Comment by quicksketch** [comment]
I added a one line of code comment to the PR just to indicate that the adding of all settings for all formats before reducing the list is intentional. Thank you @indigoxela for the test coverage!

**Comment by quicksketch** [comment]
Thank you @indigoxela, @sethhill, and @laryn! I have merged https://github.com/backdrop/backdrop/pull/4921 into 1.x and 1.29.x. This will be in 1.29.3.

Another thank you and welcome to @sethhill. Thank you for opening this report!

**Comment by quicksketch** [comment]
I'm not sure how we missed this in the test coverage, but the new test contains a syntax error that affects PHP 5.6 (other versions seem fine).

```
PHP Parse error:  syntax error, unexpected ')' in /home/runner/work/backdrop/backdrop/core/modules/filter/tests/filter_js_settings.test on line 21
ERROR: FATAL FilterSettingsAttachedTestCase: test runner returned a non-zero error code (255).
```

I've filed a follow-up PR at https://github.com/backdrop/backdrop/pull/4939.

**Comment by quicksketch** [comment]
I merged that simple change to unblock other pull requests.

**Comment by indigoxela** [comment]
> I'm not sure how we missed this in the test coverage, but the new test contains a syntax error that affects PHP 5.6...

Wow... Hm. parent::setUp is used that way in several other tests... - modules to enable not wrapped in array.
Anyway, many thanks for fixing.

**Comment by quicksketch** [comment]
The problem was actually just the trailing comma being used in the parameter list. This was the previous code:

```
    parent::setUp(
      'filter',
      'ckeditor5',
    );
```

It could have been just as easily fixed by simply removing the trailing comma:
```
    parent::setUp(
      'filter',
      'ckeditor5'
    );
```

But long-term I think we should switch a single parameter as an array for all tests, as it's easier to merge. Right now it's a total mess with some classes taking one format but not the other or classes not adhering to any of the arguments at all.

**Comment by indigoxela** [comment]
> The problem was actually just the trailing comma...

Oh, right, get it now. :see_no_evil: Many thanks for fixing. :+1:
